src/dune_engine/dep.mli

```diff
@@ -110,19 +110,28 @@ module Set : sig
   include
     Set.S with type elt = t and type 'a map := 'a Map.t and type t = unit Map.t
 
-  (** Return dependencies on all source files under a certain source directory.
-
-      Dependency on a source_tree requires special care for empty directory, so
-      you should use this function rather than manually traverse the source
-      tree. *)
-  val source_tree : Path.t -> t Memo.t
-
-  (** Same as [source_tree] but also return the set of files as a set. Because
-      of the special care for empty directories, the set of dependencies
-      returned contains dependencies other than [File]. So extracting the set of
-      files from the dependency set is a bit awkward. This is why this function
-      exist. *)
-  val source_tree_with_file_set : Path.t -> (t * Path.Set.t) Memo.t
+  module Tree
+      (Monad : Monad.S) (_ : functor (Outcome : Monoid.S) -> sig
```

Comment:
I could have fixed the monad to be Memo, but  it seems nicer to specify that we don't really need something memo specific beyond a monad here.

Reply:
Agreed that saying this works for any monad is nice!

I'm a little less sure about the functor bit. Like in the other module, the call sites might benefit from knowing that we're really using the `Union` monoid here.

Reply:
(I mean, we could switch to `Monoid.Commutative.S` but even that might be too restrictive.)

Reply:
I specialized it to Union. I agree it's a lot cleaner now.